Add Data Viz Foundation (#5709)
jpmorganchase/salt-ds: site/docs/foundations/data-visualization/chart-accessibility.mdx

---
title: Chart accessibility
layout: DetailTechnical
sidebar:
  priority: 5
---

Charts rely heavily on visuals to help users understand complex data, so it’s important to ensure that your charts meet all accessibility requirements. Use color appropriately, include visual cues, provide text for assistive technologies, and ensure any interactions are accessible.

## Categorical colors

### Color contrast

Users with visual impairments should be able to fully comprehend data sets that are represented by different colors. To cater for a broad spectrum of visual impairments, you must use the categorical colors which are WCAG 4.5:1 compliant.

> If you use custom colors or Hex codes, they may not meet non-text contrast requirements so you will be responsible to ensure that the resulting contrast is compliant. By using Salt tokens correctly, you will achieve at least 3:1 contrast between the background and other non-text chart elements such as bars, lines, segments and icons.

Reviewer: Is this something we want to even touch on? I don't think we typically talk about users using hex codes or custom colors anywhere else. Also I feel this is very simlar to the paragraph above

Author: I've adjusted the structure slightly to not repeat and shortened the para. I think it's important to mention people shouldn't choose random colors, which I've seen designers do in Figma, or at least mention they must use the right tokens.

<Image
  src="/img/foundations/dataviz-hex.png"
  srcDark="/img/foundations/dataviz-hex-dark.png"
  alt="Diagram of best practices using categorical color versus color picking."
/>

### Color order

Apply the categorical colors in the recommended order. All categorical colors have similar contrast levels, as they use the 500 series swatch from their respective color ramps.

<Image
  src="/img/foundations/dataviz-color-order.png"
  srcDark="/img/foundations/dataviz-color-order-dark.png"
  alt="Diagram of the categorical color palette."
/>

A well-defined, distinct color palette ensures that everyone can easily interpret the data. Using the set categorical palette colors minimizes the risk of misinterpretation and ensures a visual order. Don’t use shades of the same color.

<Image
  src="/img/foundations/dataviz-monochromatic.png"
  srcDark="/img/foundations/dataviz-monochromatic-dark.png"
  alt="Diagram of best practices using categorical palette versus monochromatic palette."
/>

You can enhance your chart’s accessibility by adopting additional visual strategies other than color such as:

- Pattern overlays for different categories
- Direct labelling where possible

> - Adding a border with at least a 3:1 contrast ratio to distinguish between segments

Reviewer: We should say what the 3:1 contrast ratio should be against e.g., the background of the container, the segment backgrounds or all of the above.

> You can show chart data in a table or list nearby, in an overlay, or on another screen as a last resort to support screen readers. If the essential information is in a separate screen, then it could disrupt the cohesive flow of information. Screen reader users would need to navigate back and forth between the two parts of the text alternative for the complex image, which could disrupt the flow of information.

Reviewer: This paragraph seems random under the color order heading without an intro. Also I'm not entirely sure what it's trying to say, @jake-costa are text alternatives an actual last resort or a helpful addition as it can help screen readers and really all users comprehend the information?

For more information about text alternatives, read the [JPMC UX Accessibility Resources](https://go/text-alternatives#TextAlternatives-Compleximages) (Internal users only).

### Overlapping data series

> Some chart types including area, bubble and scatter have data series that overlap in certain use cases. Having too many data series in your chart can add complexity and make it harder for users to extract important information. In cases with multiple overlapping areas, always maintain at least a 3:1 color contrast between any 2 overlapping areas.

Reviewer: ```suggestion
Some chart types, including area, bubble and scatter, have data series that overlap in certain use cases. Having too many data series in your chart can add complexity and make it harder for users to extract important information. In cases with multiple overlapping areas, always maintain at least a 3:1 color contrast between any two overlapping areas.
```

<Image
  src="/img/foundations/dataviz-overlapping-data.png"
  srcDark="/img/foundations/dataviz-overlapping-data-dark.png"
  alt="Diagram of bubble chart to show colors overlapping."
/>

## Line patterns

> Color shouldn’t be the only visual cue. Include other visual cues such as text, patterns, or symbols to help the user differentiate between different data sets. Patterns improve readability, especially for those with color vision requirements.

Author: Yes, I'd advocate to have each anatomy part as their own section

<Image
  src="/img/foundations/dataviz-line-patterns2.png"
  srcDark="/img/foundations/dataviz-line-patterns2-dark.png"
  alt="Diagram of line chart using line patterns."
/>

## Pattern fills

> Consider using line patterns for multiple line series or pattern fills for 2-dimensional shapes, such as bars and slices. Pattern fills are always black and applied to solid categorical colors for charts. They’re beneficial in scenarios in which there is only one visual distinction between data points.

Reviewer: This talks about line patterns but the section above is for line patterns

Author: Fixed

> Consider using line patterns for multiple line series or pattern fills for 2-dimensional shapes, such as bars and slices. Pattern fills are always black and applied to solid categorical colors for charts. They’re beneficial in scenarios in which there is only one visual distinction between data points.

Reviewer: ```suggestion
Consider using line patterns for multiple line series or pattern fills for 2-dimensional shapes, such as bars and slices. Pattern fills are always black and applied to solid categorical colors for charts. They’re beneficial in scenarios in which there's only one visual distinction between data points.
```

> Consider using line patterns for multiple line series or pattern fills for 2-dimensional shapes, such as bars and slices. Pattern fills are always black and applied to solid categorical colors for charts. They’re beneficial in scenarios in which there is only one visual distinction between data points.

Reviewer: For the bit about "only one visual distinction between data points", we should probably give examples because right now I could say a visual distinction between the points in the pie is the width of the segments.

<Image
  src="/img/foundations/dataviz-pattern-fills2.png"
  srcDark="/img/foundations/dataviz-pattern-fills2-dark.png"
  alt="Diagram of donut chart using pattern fills."
/>

## Marker symbols

> In addition to line patterns and pattern fills, you can add shape or icon markers to make data points more distinctive. Varied marker symbols help differentiate lines. Marker symbols have a white border in both light and dark mode for accessibility.

Reviewer: ```suggestion
In addition to line patterns and pattern fills, you can add shape or icon markers to make data points more distinctive. Varied marker symbols help differentiate lines. Marker symbols have a white border in both light and dark modes for accessibility.
```

> In addition to line patterns and pattern fills, you can add shape or icon markers to make data points more distinctive. Varied marker symbols help differentiate lines. Marker symbols have a white border in both light and dark mode for accessibility.

Reviewer: "for accessibility", we should be more specific

Reviewer: Also is this too specific to the implementation (white border)

> In addition to line patterns and pattern fills, you can add shape or icon markers to make data points more distinctive. Varied marker symbols help differentiate lines. Marker symbols have a white border in both light and dark mode for accessibility.

Reviewer: Also reading this with the diagram makes me think marker symbols only apply to line charts but scatter plots are a good use-case too

---
title: Chart accessibility
layout: DetailTechnical
sidebar:
  priority: 5
---

> Charts rely heavily on visuals to help users understand complex data, so it’s important to ensure that your charts meet all accessibility requirements. Use color appropriately, include visual cues, provide text for assistive technologies, and ensure any interactions are accessible.

Reviewer 2: I'm not sure about "Charts rely heavily on visuals". With the help of my friend LLM I've got the following suggestion, what do you think? 

> Charts rely on visual elements to make complex data easier to understand, so it’s essential to ensure they are accessible to all users. Use color thoughtfully and avoid relying on it alone to convey information. Add clear visual cues like patterns or labels, provide descriptive text for assistive technologies, and ensure interactive features are accessible by keyboard or other devices.

Author: Adjusted. Added in this new first sentence and kept the original second as this has a lower word count.

## Categorical colors

### Color contrast

Users with visual impairments should be able to fully comprehend data sets that are represented by different colors. If you use custom colors or Hex codes, they may not meet non-text contrast requirements so you will be responsible to ensure that the resulting contrast is compliant. By using Salt tokens correctly, you will achieve at least 3:1 contrast between the background and other non-text chart elements such as bars, lines, segments and icons.

> <Image

Reviewer 2: It might also be worth noting that by using `--dataviz` tokens they chart will complement other components set to display a category, such as Tag and Avatar.

> <Image

Author: I agree that we shouldn't refer to the foundational name in this case. However, since the new chart tokens aren't live, I'd like to avoid pushing this change if they don't yet exist. I'll update the image to show 'Categorical 3,' in the mean time. Some of this data viz content will need to update when the color re-order and new tokens are available.

> <Image

Author: Updated image.

  src="/img/foundations/dataviz-hex.png"
  srcDark="/img/foundations/dataviz-hex-dark.png"
  alt="Diagram of best practices using categorical color versus color picking."
/>

### Color order

Apply the categorical colors in the recommended order, from 1 through 20. 

<Image
  src="/img/foundations/dataviz-color-order.png"
  srcDark="/img/foundations/dataviz-color-order-dark.png"
  alt="Diagram of the categorical color palette."
/>

A well-defined, distinct color palette ensures that everyone can easily interpret the data. Using the set categorical palette colors minimizes the risk of misinterpretation and ensures a visual order. Don’t use shades of the same color.

> <Image

Reviewer 2: Will this be contradicted by Heatmap? Would it help to include a second 'do not do this' example, for instance, applying very similar colors to the pie chart e.g. categories 3, 9, 13, 19.

Reviewer 2: Would it also strengthen the argument if the white border was removed from the purple pie?

Reviewer 2: Could we change the repeated 'label' to say something more relevant? Apples, Pears, Bananas? Manufacturing, Retail, Hospitality?

> A well-defined, distinct color palette ensures that everyone can easily interpret the data. Using the set categorical palette colors minimizes the risk of misinterpretation and ensures a visual order. Don’t apply similar categorical color colors to the same chart.

Reviewer 3: Could this paragraph and related image be better placed on the color page. A11y requires more than just color. Recommendations about visual order (through color alone) is possibly not a strong a11y principle, but more related to the chart color principles we have provided instead.

  src="/img/foundations/dataviz-monochromatic.png"
  srcDark="/img/foundations/dataviz-monochromatic-dark.png"
  alt="Diagram of best practices using categorical palette versus monochromatic palette."
/>

You can enhance your chart’s accessibility by adopting additional visual strategies other than color such as:

- Pattern overlays for different categories
- Direct labelling where possible
- Adding a border with at least a 3:1 contrast ratio against the background to distinguish between segments

> You can show chart data in a table or list nearby, in an overlay, or on another screen as a last resort to support screen readers. If essential information is on a separate screen, then it could disrupt the cohesive flow of information. Screen reader users would need to navigate back and forth between the two parts of the text alternative for the complex image, which could disrupt the flow of information.

Reviewer 2: > 
You can present chart data in a nearby table or list, in an overlay, or—if necessary—as a last resort on a separate screen to support screen readers. However, placing essential information on a separate screen can disrupt the flow of information. Screen reader users may need to navigate back and forth between different sections to access the full text alternative for a complex image, which can make understanding the data more difficult.

Author: Adjusted, thank you!

For more information about text alternatives, read the [JPMC UX Accessibility Resources](https://go/text-alternatives#TextAlternatives-Compleximages) (Internal users only).

> ## Line patterns

Reviewer 3: Can we add links to the relevant sections in the color and pattern page.

Color shouldn’t be the only visual cue. Include other visual cues such as text, patterns, or symbols to help the user differentiate between different data sets. Consider using line patterns for multiple line series. Patterns improve readability, especially for those with color vision requirements.

> <Image

Author: Updated image.

> Color shouldn't be the only visual cue used in charts. Always include additional cues such as text, patterns, or symbols to help users distinguish between different data sets. For example, use line patterns to differentiate multiple line series. Patterns and other visual indicators improve readability, especially for users with color vision requirements.

Reviewer 3: The first half of this section reads more like an introduction to lines, patterns and markers. It's not specific to Line pattern.

  src="/img/foundations/dataviz-line-patterns2.png"
  srcDark="/img/foundations/dataviz-line-patterns2-dark.png"
  alt="Diagram of line chart using line patterns."
/>

## Pattern fills

> Use pattern fills for two-dimensional shapes, such as bars and slices. Pattern fills should always be black and applied over solid categorical colors in charts. They are especially helpful when there is only one visual distinction between data points, ensuring that all users can easily differentiate between chart elements.

Reviewer 3: Is there a reason why pattern fills should always be black? To achieve 3:1 contrast, but I assume white would be the same...

<Image
  src="/img/foundations/dataviz-pattern-fills2.png"
  srcDark="/img/foundations/dataviz-pattern-fills2-dark.png"
  alt="Diagram of donut chart using pattern fills."
/>

## Marker symbols

In addition to line patterns and pattern fills, you can add shape or icon markers to make data points more distinctive. Varied marker symbols help differentiate lines. Marker symbols have a white border in both light and dark modes for accessibility.

> By default, Salt uses predefined symbols in the following order:

Reviewer 2: Can we change this sentence so it doesn't start with 'By default'. Generally we try to avoid saying 'default'.

- Circle
- Diamond
- Square

> - Triangle

Reviewer 2: Do we need to clarify this is triangle up?

- Triangle-down

<Image
  src="/img/foundations/dataviz-markers.png"
  srcDark="/img/foundations/dataviz-markers-dark.png"
  alt="Diagram of line chart using marker symbols."
/>

> 

Author: I don’t think this is necessary, as we don’t include this in other folder sections within Foundations. To keep the word count concise, I believe consumers can navigate to the next section on their own if they want.
